src/app/recipes/recipe-edit/recipe-edit.component.ts

```diff
@@ -11,18 +12,79 @@ import { Recipe } from '../recipe.model';
 export class RecipeEditComponent implements OnInit {
   isNewRecipe = false;
   recipe: Recipe;
-  constructor(private activatedRoute: ActivatedRoute, private recipeService: RecipeService) { }
+  recipeForm: FormGroup;
+  constructor(private activatedRoute: ActivatedRoute, private recipeService: RecipeService, private router: Router) { }
 
   ngOnInit() {
     this.activatedRoute.params.subscribe(
       (params: Params) => {
         this.isNewRecipe = params['id'] == null;
-        if (this.isNewRecipe) {
-        } else {
+        if (!this.isNewRecipe) {
           this.recipe = this.recipeService.getRecipeById(parseInt(params['id'], 10));
         }
+        this.initForm();
       }
     );
   }
 
+  private initForm() {
+    let recipeName = '';
+    let recipeImagePath = '';
+    let recipeDescription = '';
+    const recipeIngredients = new FormArray([]);
+    if (!this.isNewRecipe) {
+      recipeName = this.recipe.name;
+      recipeImagePath = this.recipe.imagePath;
+      recipeDescription = this.recipe.description;
+      if (this.recipe.ingredients) {
+        for (const ingredient of this.recipe.ingredients) {
+          recipeIngredients.push(
+            new FormGroup({
+              name: new FormControl(ingredient.name, Validators.required),
+              amount: new FormControl(ingredient.amount, [Validators.required, Validators.pattern(/^[1-9]+[0-9]*$/)])
+            })
+          );
+        }
+      }
+    }
+    this.recipeForm = new FormGroup({
+      name: new FormControl(recipeName, Validators.required),
+      imagePath: new FormControl(recipeImagePath, Validators.required),
+      description: new FormControl(recipeDescription, Validators.required),
+      ingredients: recipeIngredients
+    });
+  }
+
+  onAddIngredient() {
+    (<FormArray>this.recipeForm.get('ingredients')).push(
+      new FormGroup({
+        name: new FormControl(null, Validators.required),
+        amount: new FormControl(null, [Validators.required, Validators.pattern(/^[1-9]+[0-9]*$/)])
+      }));
+  }
+
+  onSubmit() {
+    const recipe = new Recipe(
+      0,
+      this.recipeForm.value['name'],
+      this.recipeForm.value['description'],
+      this.recipeForm.value['imagePath'],
+      this.recipeForm.value['ingredients']);
+    if (this.isNewRecipe) {
+      this.recipeService.addRecipe(recipe);
+    } else {
+      recipe.id = this.recipe.id;
+      this.recipeService.updateRecipe(this.recipe.id, recipe);
+    }
+    this.onCancel();
+  }
+
+  onCancel() {
+    this.router.navigate(['../'], { relativeTo: this.activatedRoute });
+  }
+
+  onDeleteIngredient(id: number) {
+    (<FormArray>this.recipeForm.get('ingredients')).removeAt(id);
```

Comment:
Will ingredient be removed from storage(service)?

Reply:
By clicking the save button.